### PR TITLE
Examples Cleanup

### PR DESCRIPTION
**Pull request by alexander-g**
- refactored `examples/imagenet/resnet_imagenet.py` to accept parameters instead of modifying them inside the script
- added `README.md` for `examples/imagenet/`
- removed unnecessary `Lambda` class from `examples/mnist.py`
- moved global average pooling in `examples/mnist_conv.py` before the `Linear` layer

### CODE REVIEW

**Comment by alexander-g** [comment]
- Added loss scaling parameter to elegy.Model for mixed precision
- requirements_dev.txt for additional requirements for development

**Comment by cgarciae** [comment]
Hey @alexander-g, thanks for this :)

A few comments:
* I think the `loss_scaling` parameter could be avoided entirely on `Model` side and get pushed to the optimizer via `optax.scale`. 
* Thanks for cleaning up the other examples!

**Comment by alexander-g** [comment]
Loss scaling is 3 steps: 
1. Up-scale the loss
2. Compute gradients with the up-scaled loss
3. Down-scale the gradients

Only the last part can be performed with `optax.scale`. For the first part the user would additionally need to multiply the loss function(s) manually. This is 2+ locations that the user has to think of, plus the up-scaled loss would be logged in the logs, making it more difficult to compare when changing the loss scale.
So I added a parameter to `Model` to make it more user friendly.

**Comment by charlielito** [comment]
We use poetry as the package manager. So we don't need a requirements_dev.txt. Poetry handles this for you in the `--dev` dependencies. See https://github.com/poets-ai/elegy/blob/master/pyproject.toml#L32

**Comment by cgarciae** [comment]
@alexander-g I've been thinking about this and currently there are a couple of ways to go about this:

1. Using the [weight](https://poets-ai.github.io/elegy/api/Loss/#elegy.losses.loss.Loss.__init__) argument that all losses accept + `optax.scale`. 
2. Create a `Scaled(Loss)` class that handles other losses and uses [loss partitioning](https://poets-ai.github.io/elegy/guides/modules-losses-metrics/#partitioning-a-loss) to report them + `optax.scale`.

You method proposed method suggests modifying `value_and_grad` and accepting a parameter in `Model`.

I like that 1 and 2 don't require any changes to the internal and are quite general, on the other hand they require the user to be more careful. On the other hand, your method makes it simpler for the user for this specific but I wonder if its general enough, e.g. what if new methods use a more complicated loss scaling-like setup that renders this approach useless? 

I like something like the custom [LossScaleOptimizer](https://www.tensorflow.org/api_docs/python/tf/keras/mixed_precision/experimental/LossScaleOptimizer) from Keras. To be able to do this it would be required that we hand over more work to the [Optimizer](https://github.com/poets-ai/elegy/blob/3494cc7d495198f4c383d3560ea05df65bb669ff/elegy/model/model_base.py#L543) class (which wraps `optax`) such that it also performs the call to `value_and_grad` so that when we define`LossScaleOptimizer` as a subclass we can make the appropriate changes.

I have a proposal: separate the cleanup of the examples from the loss scaling issue, you can use methods 1 or 2 for now, and send a PR with just the loss scaling stuff. I will also open a PR with the proposed changes to our `Optimizer` class such the the `LossScaleOptimizer` can be implemented.

**Comment by alexander-g** [comment]
- removed loss scaling commits
- removed requirements_dev.txt
